src/main/java/io/vertx/core/Future.java

```diff
@@ -133,6 +137,38 @@
    */
   void fail(String failureMessage);
 
+  /**
+   * Compose this future with another future.
+   *
+   * When this future succeeds, the handler will be called with the value.
+   *
+   * When this future fails, the failure will be propagated to the {@code next} future.
+   *
+   * @param handler the handler
+   * @param next the next future
+   */
+  default <U> void compose(Handler<T> handler, Future<U> next) {
+    setHandler(ar -> {
+      if (ar.succeeded()) {
+        try {
+          handler.handle(ar.result());
+        } catch (Throwable err) {
+          if (next.isComplete()) {
```

Comment:
Sorry, we talked about this last week, but I'm still not sure why we should not call `next.fail(err)` if next is complete.